src/main/java/javaslang/WrappedString.java

```diff
@@ -0,0 +1,2200 @@
+    @Override
+    public Vector<WrappedString> combinations(int k) {
+        class Recursion {
+            Vector<WrappedString> combinations(WrappedString elements, int k) {
+                return (k == 0)
+                        ? Vector.of(WrappedString.empty())
+                        : elements.zipWithIndex().flatMap(t -> combinations(elements.drop(t._2 + 1), (k - 1))
+                        .map((WrappedString c) -> c.prepend(t._1)));
+            }
+        }
+        return new Recursion().combinations(this, Math.max(k, 0));
+    }
+
+    @Override
+    public WrappedString distinct() {
+        return distinctBy(Function.identity());
+    }
+
+    @Override
+    public WrappedString distinctBy(Comparator<? super Character> comparator) {
+        Objects.requireNonNull(comparator, "comparator is null");
+        final java.util.Set<Character> seen = new java.util.TreeSet<>(comparator);
+        return filter(seen::add);
+    }
+
+    @Override
+    public <U> WrappedString distinctBy(Function<? super Character, ? extends U> keyExtractor) {
+        Objects.requireNonNull(keyExtractor, "keyExtractor is null");
+        final java.util.Set<U> seen = new java.util.HashSet<>();
+        return filter(t -> seen.add(keyExtractor.apply(t)));
+    }
+
+    @Override
+    public WrappedString drop(int n) {
+        if (n >= length()) {
+            return EMPTY;
+        } if (n <= 0) {
+            return this;
+        }else {
+            return of(back.substring(n));
+        }
+    }
+
+    @Override
+    public WrappedString dropRight(int n) {
+        if (n >= length()) {
+            return EMPTY;
+        } if (n <= 0) {
+            return this;
+        }else {
+            return of(back.substring(0, length() - n));
+        }
+    }
+
+    @Override
+    public WrappedString dropWhile(Predicate<? super Character> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        int index = 0;
+        while (index < length() && predicate.test(charAt(index))) {
+            index++;
+        }
+        return index < length() ? (index == 0 ? this : of(back.substring(index))) : empty();
+    }
+
+    @Override
+    public WrappedString filter(Predicate<? super Character> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        final StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < back.length(); i++) {
+            final char ch = back.charAt(i);
+            if (predicate.test(ch)) {
+                sb.append(ch);
+            }
+        }
+        return sb.length() == 0 ? EMPTY : sb.length() == length() ? this : of(sb.toString());
+    }
+
+    @Override
+    public WrappedString findAll(Predicate<? super Character> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        return filter(predicate);
+    }
+
+    @Override
+    public  <U> Vector<U> flatMap(Function<? super Character, ? extends Iterable<? extends U>> mapper) {
```

Comment:
With `flatMap` we have a TODO in general but not here, with this pull-request.

Example:

```scala
scala> "123".flatMap(Some(_))
res0: String = 123

scala> "123".flatMap(c => Some(c + 0))
res1: scala.collection.immutable.IndexedSeq[Int] = Vector(49, 50, 51)

scala> val words = Set("Patryk", "Ruslan", "Daniel")
words: scala.collection.immutable.Set[String] = Set(Patryk, Ruslan, Daniel)

scala> words flatMap (word => word.toSet)
res2: scala.collection.immutable.Set[Char] = Set(e, s, n, y, t, u, a, i, l, P, r, R, k, D)

scala> words flatten (word => word.toSet)
res3: scala.collection.immutable.Set[Char] = Set(e, s, n, y, t, u, a, i, l, P, r, R, k, D)
```

Our flatMap works totally different. We need to check what is possible with Java, what makes sense and what is the expected behavior of flatMap. Maybe we need a Monad.unit(val) and Monoid.combine(m1, m2)...